tests/templates/test_subroutines/test_qsvt.py

```diff
@@ -621,3 +622,157 @@ def test_global_phase_not_alway_applied():
     decomposition = qml.qsvt([1], [0, 1, 2, 3, 4], wires=[0], convention="Wx").decomposition()
     for op in decomposition:
         assert not isinstance(op, qml.GlobalPhase)
+
+
+class TestRootFindingSolver:
+    @pytest.mark.parametrize(
+        "P",
+        [
+            ([0.1, 0, 0.3, 0, -0.1]),
+            ([0, 0.2, 0, 0.3]),
+            ([-0.4, 0, 0.4, 0, -0.1, 0, 0.1]),
+        ],
+    )
+    def test_complementary_polynomial(self, P):
+        """Checks that |P(z)|^2 + |Q(z)|^2 = 1 in the unit circle"""
+
+        Q = _complementary_poly(P)  # Calculate complementary polynomial Q
+
+        # Define points on the unit circle
+        theta_vals = np.linspace(0, 2 * np.pi, 100)
+        unit_circle_points = np.exp(1j * theta_vals)
+
+        for z in unit_circle_points:
+            P_val = np.polyval(P, z)
+            P_magnitude_squared = np.abs(P_val) ** 2
+
+            Q_val = np.polyval(Q, z)
+            Q_magnitude_squared = np.abs(Q_val) ** 2
+
+            assert np.isclose(P_magnitude_squared + Q_magnitude_squared, 1, atol=1e-7)
+
+    @pytest.mark.parametrize(
+        "angles",
+        [
+            ([0.1, 2, 0.3, 3, -0.1]),
+            ([0, 0.2, 1, 0.3, 4, 2.4]),
+            ([-0.4, 2, 0.4, 0, -0.1, 0, 0.1]),
+        ],
+    )
+    def test_transform_angles(self, angles):
+        """Test the transform_angles function"""
+
+        new_angles = qml.transform_angles(angles, "QSP", "QSVT")
+        assert np.allclose(angles, qml.transform_angles(new_angles, "QSVT", "QSP"))
+
+        new_angles = qml.transform_angles(angles, "QSVT", "QSP")
+        assert np.allclose(angles, qml.transform_angles(new_angles, "QSP", "QSVT"))
+
+        with pytest.raises(AssertionError, match="Invalid conversion"):
+            _ = qml.transform_angles(angles, "QFT", "QSVT")
+
+    @pytest.mark.parametrize(
+        "poly",
+        [
+            ([0.1, 0, 0.3, 0, -0.1]),
+            ([0, 0.2, 0, 0.3]),
+            ([-0.4, 0, 0.4, 0, -0.1, 0, 0.1]),
+        ],
+    )
+    def test_correctness_QSP_angles_root_finding(self, poly):
+        """Tests that angles generate desired poly"""
+
+        angles = qml.poly_to_angles(poly, "QSP", angle_solver="root-finding")
+        x = 0.5
+
+        @qml.qnode(qml.device("default.qubit"))
+        def circuit_qsp():
+            qml.RX(2 * angles[0], wires=0)
+            for angle in angles[1:]:
+                qml.RZ(-2 * np.arccos(x), wires=0)
+                qml.RX(2 * angle, wires=0)
```

Comment:
Why not test with the pennylane operation for QSP here?

Reply:
I am used to expressing qsp as a function of RZs and RXs. Both methods should work 😄